File: symfony-app/src/Service/CommentSentimentScoringService.php
<?php

namespace App\Service;

use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
use Symfony\Contracts\HttpClient\HttpClientInterface;

class CommentSentimentScoringService
{
    public function __construct(
        private readonly HttpClientInterface $httpClient
    ) {}

    /**
     * @return array{score:int,label:string,confidence:float,source:string}
     */
    public function analyze(string $text): array
    {
        $normalized = trim($text);
        if ($normalized === '') {
            return [
                'score' => 50,
                'label' => 'NEUTRAL',
                'confidence' => 0.0,
                'source' => 'empty',
            ];
        }

        $remote = $this->analyzeWithRobertaXml($normalized);
        if ($remote !== null) {
            return $remote;
        }

        return $this->fallbackAnalyze($normalized);
    }

    /**
     * Convertit le résultat d'analyse en score signé [-1, 1].
     *
     * @param array{score?:int|float,label?:string,confidence?:float} $analysis
     */
    public function toSignedScore(array $analysis): float
    {
        $label = strtoupper(trim((string) ($analysis['label'] ?? 'NEUTRAL')));
        $confidenceRaw = $analysis['confidence'] ?? null;
        $scoreRaw = $analysis['score'] ?? 0;

        if (is_numeric($confidenceRaw)) {
            $confidence = (float) $confidenceRaw;
        } elseif (is_numeric($scoreRaw)) {
            $scoreFloat = (float) $scoreRaw;
            $confidence = $scoreFloat > 1.0 ? ($scoreFloat / 100.0) : $scoreFloat;
        } else {
            $confidence = 0.0;
        }

        $confidence = max(0.0, min(1.0, $confidence));

        if ($label === 'POSITIVE') {
            return $confidence;
        }
        if ($label === 'NEGATIVE') {
            return -$confidence;
        }

        return 0.0;
    }

    /**
     * @return array{score:int,label:string,confidence:float,source:string}|null
     */
    private function analyzeWithRobertaXml(string $text): ?array
    {
        $endpoint = trim((string) ($_ENV['ROBERTA_XML_ENDPOINT'] ?? $_SERVER['ROBERTA_XML_ENDPOINT'] ?? ''));
        if ($endpoint === '') {
            return null;
        }

        $payload = [
            'text' => $text,
            'model' => (string) ($_ENV['ROBERTA_XML_MODEL'] ?? $_SERVER['ROBERTA_XML_MODEL'] ?? 'roberta-xml'),
        ];

        try {
            $response = $this->httpClient->request('POST', $endpoint, [
                'json' => $payload,
                'timeout' => 4.0,
            ]);
            $data = $response->toArray(false);
        } catch (TransportExceptionInterface|\Throwable) {
            return null;
        }

        return $this->normalizeRobertaResponse($data);
    }

    /**
     * @param mixed $data
     * @return array{score:int,label:string,confidence:float,source:string}|null
     */
    private function normalizeRobertaResponse(mixed $data): ?array
    {
        if (!is_array($data)) {
            return null;
        }

        $label = null;
        $confidence = null;
        $score = null;

        if (isset($data['score']) && is_numeric($data['score'])) {
            $score = (float) $data['score'];
        }
        if (isset($data['label']) && is_string($data['label'])) {
            $label = strtoupper(trim($data['label']));
        }
        if (isset($data['confidence']) && is_numeric($data['confidence'])) {
            $confidence = (float) $data['confidence'];
        }

        if (isset($data['result']) && is_array($data['result'])) {
            $nested = $this->normalizeRobertaResponse($data['result']);
            if ($nested !== null) {
                return $nested;
            }
        }

        if (array_is_list($data) && isset($data[0]) && is_array($data[0])) {
            $nested = $this->normalizeRobertaResponse($data[0]);
            if ($nested !== null) {
                return $nested;
            }
        }

        if ($score === null && $confidence !== null) {
            $score = $confidence <= 1.0 ? $confidence * 100.0 : $confidence;
        }

        if ($score === null || $label === null) {
            return null;
        }

        $scoreInt = (int) max(0, min(100, round($score)));
        $confidenceValue = $confidence ?? ($scoreInt / 100);
        $confidenceValue = max(0.0, min(1.0, $confidenceValue));

        return [
            'score' => $scoreInt,
            'label' => $this->normalizeLabel($label),
            'confidence' => $confidenceValue,
            'source' => 'roberta_xml',
        ];
    }

    /**
     * @return array{score:int,label:string,confidence:float,source:string}
     */
    private function fallbackAnalyze(string $text): array
    {
        $positiveWords = [
            'excellent', 'super', 'merci', 'utile', 'clair', 'parfait', 'top', 'bon', 'bien', 'aide',
            'bravo', 'genial', 'satisfait', 'recommande', 'helpful', 'great', 'love', 'nice', 'good',
        ];
        $negativeWords = [
            'nul', 'mauvais', 'horrible', 'arnaque', 'faux', 'inutile', 'déçu', 'decu', 'lent',
            'bug', 'erreur', 'grave', 'dangereux', 'haine', 'violence', 'spam', 'bad',
        ];

        $lower = mb_strtolower($text);
        $positive = 0;
        $negative = 0;

        foreach ($positiveWords as $word) {
            if (str_contains($lower, $word)) {
                $positive++;
            }
        }
        foreach ($negativeWords as $word) {
            if (str_contains($lower, $word)) {
                $negative++;
            }
        }

        $raw = 50 + ($positive * 12) - ($negative * 12);
        $score = (int) max(0, min(100, $raw));
        $label = $score >= 55 ? 'POSITIVE' : ($score <= 45 ? 'NEGATIVE' : 'NEUTRAL');
        $confidence = min(1.0, 0.45 + (abs($positive - $negative) * 0.1));

        return [
            'score' => $score,
            'label' => $label,
            'confidence' => $confidence,
            'source' => 'fallback',
        ];
    }

    private function normalizeLabel(string $label): string
    {
        if (preg_match('/label[_\s-]?2/i', $label)) {
            return 'POSITIVE';
        }
        if (preg_match('/label[_\s-]?0/i', $label)) {
            return 'NEGATIVE';
        }
        if (preg_match('/label[_\s-]?1/i', $label)) {
            return 'NEUTRAL';
        }
        if (str_contains($label, 'POS')) {
            return 'POSITIVE';
        }
        if (str_contains($label, 'NEG')) {
            return 'NEGATIVE';
        }

        return 'NEUTRAL';
    }
}

File: javafx-app/src/main/java/com/santea/service/CommentSentimentScoringService.java
package com.santea.service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Scoring sentiment commentaires : endpoint Roberta XML + fallback lexical
 * (symfony-app CommentSentimentScoringService).
 */
public class CommentSentimentScoringService {
    private static final Duration HTTP_TIMEOUT = Duration.ofSeconds(4);
    private static final Pattern LABEL_2 = Pattern.compile("label[_\\s-]?2", Pattern.CASE_INSENSITIVE);
    private static final Pattern LABEL_0 = Pattern.compile("label[_\\s-]?0", Pattern.CASE_INSENSITIVE);
    private static final Pattern LABEL_1 = Pattern.compile("label[_\\s-]?1", Pattern.CASE_INSENSITIVE);

    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(HTTP_TIMEOUT)
            .build();

    public SentimentAnalysis analyze(String text) {
        String normalized = text == null ? "" : text.trim();
        if (normalized.isEmpty()) {
            return new SentimentAnalysis(50, "NEUTRAL", Double.valueOf(0.0), "empty");
        }

        SentimentAnalysis remote = analyzeWithRobertaXml(normalized);
        if (remote != null) {
            return remote;
        }
        return fallbackAnalyze(normalized);
    }

    /**
     * Comme CommentSentimentScoringService::toSignedScore (PHP) : confidence prioritaire, sinon score.
     */
    public double toSignedScore(SentimentAnalysis analysis) {
        if (analysis == null) {
            return 0.0;
        }
        String label = analysis.label() == null ? "NEUTRAL" : analysis.label().toUpperCase(Locale.ROOT).trim();

        double confidence;
        if (analysis.confidence() != null) {
            confidence = analysis.confidence();
        } else {
            confidence = deriveConfidenceFromScore(analysis.score());
        }

        confidence = Math.max(0.0, Math.min(1.0, confidence));

        if ("POSITIVE".equals(label)) {
            return confidence;
        }
        if ("NEGATIVE".equals(label)) {
            return -confidence;
        }
        return 0.0;
    }

    private static double deriveConfidenceFromScore(double scoreRaw) {
        double scoreFloat = scoreRaw;
        return scoreFloat > 1.0 ? (scoreFloat / 100.0) : scoreFloat;
    }

    private SentimentAnalysis analyzeWithRobertaXml(String text) {
        String endpoint = getenv("ROBERTA_XML_ENDPOINT");
        if (endpoint.isEmpty()) {
            return null;
        }

        String model = getenv("ROBERTA_XML_MODEL");
        if (model.isEmpty()) {
            model = "roberta-xml";
        }

        String body = CommunityHttpJson.buildTextModelBody(text, model);
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(endpoint))
                    .timeout(HTTP_TIMEOUT)
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                    .build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                return null;
            }
            return normalizeRobertaResponse(response.body());
        } catch (Exception e) {
            return null;
        }
    }

    private SentimentAnalysis normalizeRobertaResponse(String raw) {
        if (raw == null || raw.isBlank()) {
            return null;
        }
        String slice = CommunityHttpJson.unwrapNestedResponse(raw);

        Double score = CommunityHttpJson.findNumericField(slice, "score");
        String labelStr = CommunityHttpJson.findStringField(slice, "label");
        Double confidence = CommunityHttpJson.findNumericField(slice, "confidence");

        if (score == null && confidence != null) {
            score = confidence <= 1.0 ? confidence * 100.0 : confidence;
        }

        if (score == null || labelStr == null) {
            return null;
        }

        int scoreInt = (int) Math.max(0, Math.min(100, Math.round(score)));
        double confidenceValue = confidence != null ? confidence : (scoreInt / 100.0);
        confidenceValue = Math.max(0.0, Math.min(1.0, confidenceValue));

        return new SentimentAnalysis(scoreInt, normalizeLabel(labelStr), Double.valueOf(confidenceValue), "roberta_xml");
    }

    private SentimentAnalysis fallbackAnalyze(String text) {
        String[] positiveWords = {
                "excellent", "super", "merci", "utile", "clair", "parfait", "top", "bon", "bien", "aide",
                "bravo", "genial", "satisfait", "recommande", "helpful", "great", "love", "nice", "good",
        };
        String[] negativeWords = {
                "nul", "mauvais", "horrible", "arnaque", "faux", "inutile", "déçu", "decu", "lent",
                "bug", "erreur", "grave", "dangereux", "haine", "violence", "spam", "bad",
        };

        String lower = text.toLowerCase(Locale.ROOT);
        int positive = 0;
        int negative = 0;

        for (String word : positiveWords) {
            if (lower.contains(word)) {
                positive++;
            }
        }
        for (String word : negativeWords) {
            if (lower.contains(word)) {
                negative++;
            }
        }

        int raw = 50 + (positive * 12) - (negative * 12);
        int score = Math.max(0, Math.min(100, raw));
        String label = score >= 55 ? "POSITIVE" : (score <= 45 ? "NEGATIVE" : "NEUTRAL");
        double conf = Math.min(1.0, 0.45 + (Math.abs(positive - negative) * 0.1));

        return new SentimentAnalysis(score, label, Double.valueOf(conf), "fallback");
    }

    private String normalizeLabel(String label) {
        if (label == null) {
            return "NEUTRAL";
        }
        if (LABEL_2.matcher(label).find()) {
            return "POSITIVE";
        }
        if (LABEL_0.matcher(label).find()) {
            return "NEGATIVE";
        }
        if (LABEL_1.matcher(label).find()) {
            return "NEUTRAL";
        }
        String u = label.toUpperCase(Locale.ROOT);
        if (u.contains("POS")) {
            return "POSITIVE";
        }
        if (u.contains("NEG")) {
            return "NEGATIVE";
        }
        return "NEUTRAL";
    }

    private static String getenv(String key) {
        String v = System.getenv(key);
        return v == null ? "" : v.trim();
    }

    public record SentimentAnalysis(int score, String label, Double confidence, String source) {
    }
}
